Add user profile capabilities

darryl-sw opened the pull request:
This adds the ability for users to visually maintain their session on the frontend, as well as to add and remove categories that they may be interested in. Users are not able to view or add purchase history yet - this requires further discussion with the team (checkout process).

darryl-sw committed:
Add Profile Page

Content of profile does not generate dynamically. Requires fixing.

darryl-sw committed:
Modify Profile Page

Made use of local storage in vue expression. Got rid of redundant dependencies.

darryl-sw committed:
Update profileComponent.js

Frontend for categories

darryl-sw committed:
Add ability to add categories as user

darryl-sw committed:
Resolve Bugs in Users

Make categories blueprint non-mandatory to support users that have not yet selected any categories.

Bugs to solve: error prompt when adding the first category for user.

darryl-sw committed:
Add ability to remove categories as user

Does basic validation check at frontend to ensure that calls to server are not done if categories already exist.

Resolved add categories bug.

WynWinz commented on api/users/usersController.js:
```diff
@@ -42,5 +43,46 @@ module.exports.factory = (
       })
   })
 
+  router.put('/users/:email/addcat', function (req, res) {
```
In REST URIs are a nice representation of a resource. Your current format isn't RESTful as it includes 'addcat' and 'remcat' The add category should be a PUT as you listed. I would suggest the URI being '/users/:email/category' This is now saying update the category resource at email (unique ID).

darryl-sw replied:
👍

WynWinz commented on api/users/usersController.js:
```diff
@@ -42,5 +43,46 @@ module.exports.factory = (
+      })
+  })
+
+  router.put('/users/:email/remcat', function (req, res) {
```
Similar suggestion. Depending on how you want to implement it, you can restructure and make this a DELETE instead of a PUT. https://stackoverflow.com/questions/37138720/restful-api-delete-some-of-a-resource-collection This link shows how you could specify a DELETE and delete one category such as DELETE '/users/:email/category?category=fiction' This would say delete fiction from the user's list of categories

WynWinz replied:
With this delete you will no longer need a body. You will instead be getting the category from the request params.

darryl-sw replied:
👍

WynWinz commented on web/scripts/users/loginComponent.js:
```diff
@@ -35,6 +35,11 @@ module.exports = {
             storage.set('user', res.user)
+
+            // Set username view
+            document.getElementById("username-view").innerHTML
```
Please resolve the eslint errors. VS Code should be underlining in red what they are.

darryl-sw replied:
👍  Looks like there was a plugin issue on my VS Code. It's been resolved.

WynWinz commented on web/scripts/users/profileComponent.js:
```diff
@@ -0,0 +1,102 @@
+          usersRepo.remCat(user.email, cat, (err, res) => {
+            if (err) {
+              alert('Unable to remove category: check connection')
+              return
```
There are eslint errors in this file as well.

darryl-sw replied:
👍

WynWinz commented on web/scripts/users/registrationComponent.js:
```diff
@@ -39,6 +39,11 @@ module.exports = {
+
+            // Set username view
+            document.getElementById("username-view").innerHTML
+                = "<u>" + res.user.name + "</u>"
```
Single quote only eslint errors.

darryl-sw replied:
👍

WynWinz commented on web/scripts/users/usersRepo.js:
```diff
@@ -21,6 +21,20 @@ module.exports = {
-    return { login, register }
+    const addCat = (email, categories, callback) => {
+      repo.put({
+        path: '/users/' + email + '/addcat',
```
As I mentioned on the API side, I would suggest making the same changes here with the URI.

darryl-sw replied:
👍

WynWinz commented on web/scripts/users/usersRepo.js:
```diff
@@ -21,6 +21,20 @@ module.exports = {
+    }
+
+    const remCat = (email, categories, callback) => {
+      repo.put({
```
See PUT to DELETE comment from API side.

darryl-sw replied:
👍

WynWinz commented on web/styles/main.css:
```diff
@@ -1 +1,15 @@
\ No newline at end of file
+@import url(http://fonts.googleapis.com/css?family=Lato:300,400,700,900);html{height:100%}body{font-family:'Lato',Calibri,Arial,sans-serif;background:#f9f9f9;font-weight:300;font-size:15px;color:#333;overflow:scroll;overflow-x:hidden}a{color:#555;text-decoration:none}a:hover{color:#EA2E49;text-decoration:none}.container{width:100%;position:absolute;top:120px;bottom:0;left:0;right:0}.main{position ...
+
+.td{
```
Main.css is an auto-generated file when using less. http://lesscss.org/usage/ Did you manually add the .td and .form-control-sm? Please add it in I would assume the _profile.less file. Then when you compile using lessc main.less main.css all of the CSS in the less files will be included.

WynWinz replied:
I think the best way to stop us from having issues with Main.css from now on is to add it to the .gitignore. Can you please do that in your next commit. We will mention in the README.md for the user to run npm less to fix the styles. In the web/package.json we would need "less": "lessc styles/main.less styles/main.css" added to the scripts.

darryl-sw replied:
I tried using less to no avail. Let's resolve this problem at a later time since it's a styling issue.

darryl-sw committed:
Corrected REST URI and bugs

REST URIs now follows convention. In addition, bug for adding category has been resolved. Code also follows eslint rules.

darryl-sw committed:
Minor coding style changes

darryl-sw committed:
Update .gitignore

WynWinz commented on .gitignore:
```diff
@@ -57,3 +57,8 @@ typings/
+*.json
+
+invoke-rest.ps1
+web/styles/main.less
```
Please remove main.less from .gitignore. We only want main.css added here.

darryl-sw replied:
👍

darryl-sw committed:
Update .gitignore

Allowed main.less to be committed.

darryl-sw committed:
Update .gitignore

Allows JSON files

WynWinz committed:
Merge branch 'master' into users
